### DISCUSSION TITLE
Is Parser.js doing too much?

### DISCUSSION THREAD

**Discussion by calculuschild** [markedjs/marked#2031] [Ideas]
At a certain point, splitting the functionality of Marked.js across multiple documents becomes harder and harder to track and maintain. Especially when the flow of `source -> token -> html` gets spread across several sections of code. Just looking over `parser.js`, it looks like mainly it's just a large switch statement that concatenates the output HTML string by calling `renderers` in the right order. That's fine.

Only, there's also large amounts of processing in `parser.js` for `Table` and `List` tokens which seems out of place. Would it make sense to move that processing into some larger `renderer`s? Instead of the Table renderer just wrapping the Header and Body from `parser.js` in `<table>` tags? I suspect the speed would remain similar, while also opening up more of that background processing on the `Renderer.js` side for those who want to extend Tables, for instance.

It seems like the core functions should be delineated as follows:

* Rules.js -> Regex to detect tokens in the source
* Tokenizer.js -> Test rules and process matches into Token objects (including sub-tokens)
* Lexer.js -> Determine order of Tokenizer execution, and collect Tokens into an ordered list
* Parser.js -> Determine top-level order of Renderer execution, and concatenate HTML into a single output string
* Renderer.js -> Process Token objects into HTML (including its own sub-tokens)

**Comment by UziTech** [comment]
Are you proposing that we remove the `tablerow` and `tablecell` renderers and just do it all in the `table` renderer?

**Reply by UziTech** [reply]
I feel like that would make the renderers a lot more complicated. One could make the argument to do the opposite and take the highlighting out of the code renderer and put it into the parser, for example.

**Reply by calculuschild** [reply]
No, I think sub-renderers make sense to keep. Just move the logic from the parser into a Table renderer, which then calls the table cell and table row renderers internally as needed. Rather than having program flow jumping and out of parser.js and renderer.js to handle each component of a table.

**Reply by calculuschild** [reply]
A user looking to extend the overall flow from Table token to HTML has a lot of that process locked away in the parser.

**Reply by UziTech** [reply]
A user who wants to add a class to each table would have to copy and maintain a lot of code. Including handling the inline tokens or sending those tokens back to the parser. That doesn't sound like a better user experience to me.

**Reply by calculuschild** [reply]
Well, we already have users needing to copy and maintain large blocks of code for any of the Tokenizers, for example, so it doesn't make sense to me to make an exception here.

But that's not the point anyway. Needing to copy ~32 lines of code to enable customizing the table renderer would be a better user experience in my mind than not being able to customize it at all.

**Reply by UziTech** [reply]
The parser is supposed to handle the parts of rendering the tokens that the user doesn't need to change. I don't see a reason the user would ever want to do something different with the table token. And if they do that is what `walkTokens` is for.

**Reply by calculuschild** [reply]
Well, supposedly the user doesn't need to change *any* part of marked, but that's the point of supporting extensions. So they can change things.

**Reply by UziTech** [reply]
What is there to customize in the table parser? It essentially just loops through the child tokens and parses each of them.

**Reply by calculuschild** [reply]
So, this might not be the best example, but there are some [Markdown variants](http://fletcher.github.io/MultiMarkdown-4/tables) that support column span/row span by changing up the syntax. Getting this to work with Marked.js might require looking further back in the Tokenizer or use walkTokens or something, but I could see and have seen related syntax where the contents of a cell (maybe the header has some symbol) will determine how the rest of the table is laid out. Having access to the sub-tokens of a Table token (the contents of a TableCell maybe) would be required to determine the presence of specified parameters (column width, etc.) in the resulting Table. What I'm trying to say is I could see cases where full knowledge of the Table token tree would be helpful in determining the render process.

**Reply by calculuschild** [reply]
But mostly, use cases aside, its just that having half of the logic separated out into a different file but for only a couple of cases feels weird to me. I am of course the newcomer here, so you guys have the final say on anything. Just pointing out some weirdness that maybe I'm not understanding fully.

**Reply by UziTech** [reply]
It may feel weird but there doesn't seem to be a better way to do it without requiring the user to handle the parsing logic in a render function.
